checker/src/main/java/org/checkerframework/checker/nullness/SystemGetPropertyHandler.java

```diff
@@ -71,27 +91,78 @@
         this.env = env;
         this.factory = factory;
+        this.checker = checker;
 
         systemGetProperty =
                 TreeUtils.getMethod(java.lang.System.class.getName(), "getProperty", 1, env);
+        systemSetProperty =
+                TreeUtils.getMethod(java.lang.System.class.getName(), "setProperty", 2, env);
+        systemClearProperty =
+                TreeUtils.getMethod(java.lang.System.class.getName(), "clearProperty", 1, env);
+        systemSetProperties =
+                TreeUtils.getMethod(java.lang.System.class.getName(), "setProperties", 1, env);
     }
 
+    /**
+     * If the first argument of a method call is a literal, return it; otherwise return null.
+     *
+     * @param tree a method invocation whose first formal parameter is of String type
+     * @return the first argument if it is a literal, otherwise null
+     */
+    private static @Nullable String literalFirstArgument(MethodInvocationTree tree) {
+        List<? extends ExpressionTree> args = tree.getArguments();
+        assert args.size() > 0;
+        ExpressionTree arg = args.get(0);
+        if (arg.getKind() == Tree.Kind.STRING_LITERAL) {
+            String literal = (String) ((LiteralTree) arg).getValue();
+            return literal;
+        }
+        return null;
+    }
+
+    /**
+     * Apply rules regarding System.getProperty and related methods.
+     *
+     * @param tree a method invocation
+     * @param method the method being invoked
+     */
     public void handle(MethodInvocationTree tree, AnnotatedExecutableType method) {
-        if (TreeUtils.isMethodInvocation(tree, systemGetProperty, env)) {
-            List<? extends ExpressionTree> args = tree.getArguments();
-            assert args.size() == 1;
-            ExpressionTree arg = args.get(0);
-            if (arg.getKind() == Tree.Kind.STRING_LITERAL) {
-                String literal = (String) ((LiteralTree) arg).getValue();
-                if (predefinedSystemProperties.contains(literal)) {
-                    AnnotatedTypeMirror type = method.getReturnType();
-                    type.replaceAnnotation(factory.NONNULL);
-                }
+        if (permitClearProperty) {
+            return;
+        }
+        if (TreeUtils.isMethodInvocation(tree, systemGetProperty, env)
+                || TreeUtils.isMethodInvocation(tree, systemSetProperty, env)) {
+            String literal = literalFirstArgument(tree);
+            if (literal != null && predefinedSystemProperties.contains(literal)) {
+                AnnotatedTypeMirror type = method.getReturnType();
+                type.replaceAnnotation(factory.NONNULL);
             }
         }
+        if (TreeUtils.isMethodInvocation(tree, systemClearProperty, env)) {
+            String literal = literalFirstArgument(tree);
+            if (literal == null || predefinedSystemProperties.contains(literal)) {
+                checker.reportError(tree, "clear.system.property");
```

Comment:
Errors should be reported in the visitor rather than the annotated type factory.